saexec/execution.go

```diff
@@ -0,0 +1,193 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package saexec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"runtime"
+	"time"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core"
+	"github.com/ava-labs/libevm/core/state"
+	"github.com/ava-labs/libevm/core/state/snapshot"
+	"github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/core/vm"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/strevm/blocks"
+	"github.com/ava-labs/strevm/dummy"
+	"github.com/ava-labs/strevm/hook"
+)
+
+var errExecutorClosed = errors.New("saexec.Executor closed")
+
+// Enqueue pushes a new block to the FIFO queue. If [Executor.Close] is called
+// before [blocks.Block.Executed] returns true then there is no guarantee that
+// the block will be executed.
+func (e *Executor) Enqueue(ctx context.Context, block *blocks.Block) error {
+	for {
+		select {
+		case e.queue <- block:
+			return nil
+		case <-e.quit:
+			return errExecutorClosed
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// If this happens then increase the channel's buffer size.
+			e.log.Warn("Execution queue buffer too small")
+			runtime.Gosched()
+		}
+	}
+}
+
+func (e *Executor) processQueue() {
+	defer close(e.done)
+
+	for {
+		select {
+		case <-e.quit:
+			return
+
+		case block := <-e.queue:
+			logger := e.log.With(
+				zap.Uint64("block_height", block.Height()),
+				zap.Uint64("block_time", block.BuildTime()),
+				zap.Any("block_hash", block.Hash()),
```

Comment:
I try to avoid `zap.Any`.

```suggestion
				zap.Stringer("block_hash", block.Hash()),
```

There is almost always a better type to use.